File: quickstart/src/ws.ts
import Nile from "@theniledev/js";

var emoji = require('node-emoji');

import * as dotenv from 'dotenv';

dotenv.config({ override: true })

// Get Nile URL and workspace
let envParams = [
  "NILE_URL",
  "NILE_WORKSPACE",
  "NILE_DEVELOPER_EMAIL",
  "NILE_DEVELOPER_PASSWORD",
]
envParams.forEach( (key: string) => {
  if (!process.env[key]) {
    console.error(emoji.get('x'), `Error: missing environment variable ${ key }. See .env.defaults for more info and copy it to .env with your values`);
    process.exit(1);
  }
});
const NILE_URL = process.env.NILE_URL!;
const NILE_WORKSPACE = process.env.NILE_WORKSPACE!;
const NILE_DEVELOPER_EMAIL = process.env.NILE_DEVELOPER_EMAIL!;
const NILE_DEVELOPER_PASSWORD = process.env.NILE_DEVELOPER_PASSWORD!;
const nile!;

async function loginDeveloper() {

  if (!process.env.NILE_DEVELOPER_EMAIL || !process.env.NILE_DEVELOPER_PASSWORD) {
    console.error(emoji.get('x'), `Error: please provide NILE_DEVELOPER_EMAIL and NILE_DEVELOPER_PASSWORD in .env .  See .env.defaults for more info and copy it to .env with your values`);
    process.exit(1);
  }

  nile = await Nile({
    basePath: NILE_URL,
    workspace: NILE_WORKSPACE,
  });

  await nile.developers.loginDeveloper({
     loginInfo: {
       email: NILE_DEVELOPER_EMAIL,
       password: NILE_DEVELOPER_PASSWORD,
     },
     }).catch((error:any) => {
       console.error(emoji.get('x'), `Error: Failed to login to Nile as developer ${NILE_DEVELOPER_EMAIL}: ` + error.message);
       process.exit(1);
     });

  nile.authToken = nile.developers.authToken;
  if (nile.authToken) {
    console.log("\n" + emoji.get('arrow_right'), ` Logged into Nile as developer`);
  } else {
    console.error(emoji.get('x'), ` Could not log into Nile.  Did you follow the instructions in https://github.com/TheNileDev/examples/blob/main/README.md#setup , create your Nile workspace, and set valid parameter values in a local .env file?`);
    process.exit(1);
  }

}

async function createWorkspace() {

  await loginDeveloper();

  // Check if workspace exists, create if not
  var myWorkspaces = await nile.workspaces.listWorkspaces()
  if ( myWorkspaces.find( ws => ws.name==NILE_WORKSPACE) != null) {
         console.log(emoji.get('dart'), "Workspace " + NILE_WORKSPACE + " exists");
  } else {
      await nile.workspaces.createWorkspace({
        createWorkspaceRequest: { name: NILE_WORKSPACE },
      }).then( (ws) => { if (ws != null)  console.log(emoji.get('white_check_mark'), "Created workspace: " + ws.name)})
        .catch((error:any) => {
          if (error.message == "workspace already exists") {
            console.error(emoji.get('x'), `Error: workspace ${NILE_WORKSPACE} already exists (workspace names are globally unique)`);
          } else {
            console.error(error);
            process.exit(1);
          }
        });
  }

}

createWorkspace();

File: quickstart-python/src/ws.py
#! /usr/bin/env python

import os
import sys

from dotenv import load_dotenv
from emoji import emojize

from nile_api import AuthenticatedClient, Client
from nile_api.api.developers import login_developer
from nile_api.api.workspaces import create_workspace, list_workspaces
from nile_api.models.login_info import LoginInfo
from nile_api.models.create_workspace_request import CreateWorkspaceRequest

GOOD = emojize(":check_mark_button:")
BAD = emojize(":red_circle:")
DART = emojize(":dart:", language="alias")
ARROW_RIGHT = emojize(":arrow_right:", language="alias")

load_dotenv(override=True)
params = {
    param: os.environ.get(param)
    for param in [
        "NILE_URL",
        "NILE_WORKSPACE",
        "NILE_DEVELOPER_EMAIL",
        "NILE_DEVELOPER_PASSWORD",
    ]
}

for name, value in params.items():
    if not value:
        print(
            f"{BAD} Error: missing environment variable {name}. See .env.defaults for more info and copy it to .env with your values",  # noqa: E501
        )
        sys.exit(1)


def run():
    response = login_developer.sync(
        client=Client(base_url=params["NILE_URL"]),
        info=LoginInfo(
            email=params["NILE_DEVELOPER_EMAIL"],
            password=params["NILE_DEVELOPER_PASSWORD"],
        ),
    )
    token = getattr(response, "token", None)
    if token is None:
        print(f"{BAD} Error: Failed to login to Nile as developer {params['NILE_DEVELOPER_EMAIL']}: {response.message}")
        sys.exit(1)

    client = AuthenticatedClient(base_url=params["NILE_URL"], token=token)

    print(f"\n{ARROW_RIGHT} Logged into Nile as developer {params['NILE_DEVELOPER_EMAIL']}")
    print(f"export NILE_ACCESS_TOKEN={token}")

    workspaces = list_workspaces.sync(client=client)
    workspace = next(
        (each for each in workspaces if each.name == params["NILE_WORKSPACE"]),
        None,
    )
    if workspace is not None:
        print(f"{DART} Workspace {workspace.name!r} exists")
    else:
        response = create_workspace.sync_detailed(
            client=client,
            json_body=CreateWorkspaceRequest(name=params["NILE_WORKSPACE"]),
        )
        if response.parsed is None:
            print(f"{BAD} Workspace {params['NILE_WORKSPACE']!r} already exists (workspace names are globally unique)")
            sys.exit(1)
        workspace = response.parsed
        print(f"{GOOD} Created workspace {workspace.name!r}")


if __name__ == "__main__":
    run()
